src/main/java/net/neoforged/neoforge/common/util/AttributeUtil.java

```diff
@@ -0,0 +1,356 @@
+     */
+    public static final Comparator<AttributeModifier> ATTRIBUTE_MODIFIER_COMPARATOR = Comparator.comparing(AttributeModifier::operation)
+            .thenComparing(Comparator.comparingDouble(a -> -Math.abs(a.amount()))) // Sort most impactful modifiers first
+            .thenComparing(Comparator.comparing(AttributeModifier::id));
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    /**
+     * Checks if attribute modifier tooltips should show, and if they should, adds tooltips for all attribute modifiers present on an item stack to the stack's tooltip lines.
+     * <p>
+     * After the tooltip lines have been added, fires the {@link AddAttributeTooltipsEvent} to allow mods to add additional attribute-related lines.
+     * 
+     * @param tooltip A consumer to add the tooltip lines to.
+     * @param ctx     The tooltip context.
+     */
+    public static void addAttributeTooltips(ItemStack stack, Consumer<Component> tooltip, AttributeTooltipContext ctx) {
+        ItemAttributeModifiers modifiers = stack.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
+        if (modifiers.showInTooltip()) {
+            AttributeUtil.applyModifierTooltips(stack, tooltip, ctx);
+        }
+        NeoForge.EVENT_BUS.post(new AddAttributeTooltipsEvent(stack, tooltip, ctx));
+    }
+
+    /**
+     * Applies the attribute modifier tooltips for all attribute modifiers present on the item stack.
+     * <p>
+     * Before application, this method posts the {@link GatherSkippedAttributeTooltipsEvent} to determine which tooltips should be skipped.
+     * <p>
+     * This method is also responsible for adding the modifier group category labels.
+     * 
+     * @param tooltip A consumer to add the tooltip lines to.
+     * @param ctx     The tooltip context.
+     */
+    public static void applyModifierTooltips(ItemStack stack, Consumer<Component> tooltip, AttributeTooltipContext ctx) {
+        var event = NeoForge.EVENT_BUS.post(new GatherSkippedAttributeTooltipsEvent(stack, ctx));
+        if (event.isSkippingAll()) {
+            return;
+        }
+
+        for (EquipmentSlotGroup group : EquipmentSlotGroup.values()) {
+            if (event.isSkipped(group)) {
+                continue;
+            }
+
+            Multimap<Holder<Attribute>, AttributeModifier> modifiers = getSortedModifiers(stack, group);
+
+            // Remove any skipped modifiers before doing any logic
+            modifiers.values().removeIf(m -> event.isSkipped(m.id()));
+
+            if (modifiers.isEmpty()) {
+                continue;
+            }
+
+            // Add an empty line, then the name of the group, then the modifiers.
+            tooltip.accept(Component.empty());
+            tooltip.accept(Component.translatable("item.modifiers." + group.getSerializedName()).withStyle(ChatFormatting.GRAY));
+
+            AttributeUtil.applyTextFor(stack, tooltip, modifiers, ctx);
```

Comment:
We are already in this class, we don't need to explicitly reference `AttributeUtil`